### PR TITLE
feat: Voodoo conversations types and roundtrip over network

### PR DESCRIPTION
**Pull request by michaelx11**
## Overview

This PR builds out complete Conversation functionality over network for `Voodoo` in xmtp-js. This means contact "bundles" are uploaded, invites are sent out, and conversations are populated from convo topics.

Notably: this does not store the user's VoodooIdentity anywhere. This can be problematic with xmtp-inbox-web since it reinitializes the Client three or more times. I hit this while testing.

## Changes

- New types: VoodooMessage, EncryptedVoodooMessage (goes on wire), etc
- Renamed inbound/outbound to `newVoodooInvite` and `processVoodooInvite` for significantly more clarity
- `src/voodoo/conversations` - VoodooConversation and VoodooConversations are types that manage sessionId's and locally cache messages

## Test Plan
- Roundtrip conversation tests

## Next Steps

Hook it up to a client!

### CODE REVIEW

File: src/voodoo/conversations/VoodooConversations.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+      // Check for new conversations in our invite topic
+      const inviteTopic = buildVoodooUserInviteTopic(this.client.address)
+      const newConvosFromInvites = await this.client.listEnvelopes(
+        inviteTopic,
+        this.processInvite.bind(this)
+      )
+
+      for (const convo of newConvosFromInvites) {
+        if (convo) {
+          this.conversations.set(convo.peerAddress, convo)
```
Comment (neekolas) [inline src/voodoo/conversations/VoodooConversations.ts:38]: This'll break once we introduce `conversationId` support, but fine for now

File: test/voodoo/conversations/VoodooConversation.test.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      // Check that invite is processed by bob
+      const bcs = await bob.conversations.list()
+      expect(bcs).toHaveLength(1)
+
+      const bc = bcs[0]
+
+      // Alice sends a message
+      await ac.send('hi')
+      // This should show up in alice's inbox
+      const ams = await ac.messages()
```
Comment (neekolas) [inline test/voodoo/conversations/VoodooConversation.test.ts:55]: 💥

File: test/voodoo/conversations/VoodooConversation.test.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+  describe('voodoo', () => {
+    beforeEach(async () => {
+      alice = await newLocalHostVoodooClient()
+      bob = await newLocalHostVoodooClient()
+      await waitForUserContact(alice, alice)
+      await waitForUserContact(bob, bob)
+    })
+
+    it('v2 conversation', async () => {
```
Comment (neekolas) [inline test/voodoo/conversations/VoodooConversation.test.ts:33]: Aren't we on v3 now?

**Review by neekolas — APPROVED** [review]
This is huge progress

File: src/voodoo/conversations/VoodooConversation.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        console.log('Failed to decrypt message', e)
+      }
+    })
+
+    // Wait for all promises in decryptedMessages to resolve
+    const results = await Promise.all(decryptedMessages)
+
+    for (const m of results) {
+      if (m) {
+        this._messages.push(m)
```
Comment (neekolas) [inline src/voodoo/conversations/VoodooConversation.ts:70]: Post delivery storage ✅

File: test/voodoo/conversations/VoodooConversation.test.ts (hunk 3 above)
Comment (michaelx11) [reply test/voodoo/conversations/VoodooConversation.test.ts:33]: good catch 😅

File: src/voodoo/conversations/VoodooConversations.ts (hunk 1 above)
Comment (michaelx11) [reply src/voodoo/conversations/VoodooConversations.ts:38]: I'll make sure to call it out with a comment

**Pull request merged**